t/patches/01-resolve-hosts-files.t

```diff
@@ -0,0 +1,165 @@
+# vim:set ft= ts=4 sw=4 et fdm=marker:
+use Test::Nginx::Socket::Lua;
+use Cwd qw(cwd);
+
+plan tests => blocks() * repeat_each() * 3 + 2;
+
+our $CWD = cwd();
+
+$ENV{TEST_NGINX_RESOLVER} ||= '8.8.8.8';
+$ENV{TEST_NGINX_HOSTS_FILE_PATH} = "$::CWD/t/fixtures/hosts";
+
+run_tests()
+
+__DATA__
+
+=== TEST 1: does not resolve localhost without hosts_file
+--- config
+    resolver $TEST_NGINX_RESOLVER;
+
+    location = /echo {
+        echo "Hello";
+    }
+
+    location = /t {
+        set $upstream localhost;
+        proxy_pass http://$upstream:$TEST_NGINX_SERVER_PORT/echo;
+    }
+--- request
+GET /t
+--- error_code: 502
+--- error_log
+localhost could not be resolved (3: Host not found)
+
+
+
+=== TEST 2: hosts_file option parses hosts file
+--- config
+    resolver $TEST_NGINX_RESOLVER hosts_file=$TEST_NGINX_HOSTS_FILE_PATH;
+
+    location = /t {
+        echo "Hello";
+    }
+--- request
+GET /t
+--- error_log eval
+[
+    qr/\[notice\] .*? host "localhost" will resolve to "127\.0\.0\.1"/,
+    qr/\[notice\] .*? host "localhost\.dev" will resolve to "127\.0\.0\.1"/,
+    qr/\[notice\] .*? host "localhost-v6" will resolve to "::1"/,
+    qr/\[notice\] .*? host "localhost2" will resolve to "127\.0\.0\.2"/,
+]
+--- no_error_log
+[error]
+host "ignoreme" will resolve to "172.168.0.1"
+
+
+
+=== TEST 3: hosts_file option caches parsed hostnames
+--- config
+    resolver $TEST_NGINX_RESOLVER hosts_file=$TEST_NGINX_HOSTS_FILE_PATH;
+
+    location = /echo {
+        echo "Hello";
+    }
+
+    location = /t {
+        set $upstream localhost;
+        proxy_pass http://$upstream:$TEST_NGINX_SERVER_PORT/echo;
+    }
+--- request
+GET /t
+--- response_body
+Hello
+--- no_error_log
+[error]
+
+
+
+=== TEST 4: hosts_file option caches parsed hostnames (bis)
+--- config
+    resolver $TEST_NGINX_RESOLVER hosts_file=$TEST_NGINX_HOSTS_FILE_PATH;
+
+    location = /echo {
+        echo "Hello";
+    }
+
+    location = /t {
+        set $upstream localhost.dev;
+        proxy_pass http://$upstream:$TEST_NGINX_SERVER_PORT/echo;
+    }
+--- request
+GET /t
+--- response_body
+Hello
+--- no_error_log
+[error]
+
+
+
+=== TEST 5: hosts_file option honors ipv6=off flag
+--- config
+    resolver $TEST_NGINX_RESOLVER hosts_file=$TEST_NGINX_HOSTS_FILE_PATH ipv6=off;
+
+    location = /t {
+        echo "Hello";
+    }
+--- request
+GET /t
+--- response_body
+Hello
+--- no_error_log
+host "localhost-v6" will resolve to "::1"
```

Comment:
This test will fail when Nginx is built without ipv6 for the reason highlighted above for now.